File: DAA/Python Implementation/Week9/1.py
import sys

def tsp(graph, s):
    n = len(graph)
    visited = [False] * n
    visited[s] = True
    return tsp_helper(graph, s, visited)

def tsp_helper(graph, s, visited):
    n = len(graph)
    if all(visited):
        return graph[s][0]
    min_cost = sys.maxsize
    for i in range(n):
        if not visited[i]:
            visited[i] = True
            cost = graph[s][i] + tsp_helper(graph, i, visited)
            min_cost = min(min_cost, cost)
            visited[i] = False
    return min_cost

graph = [[0, 20, 18, 20], [20, 0, 18,12], [18, 18, 0, 6], [10, 12, 6, 0]]
print("Minimum weight Hamiltonian Circuit:", tsp(graph, 0))

File: DAA/C Implementation/Week9/1.c
#include <stdio.h>
#include <limits.h>

#define V 4

int graph[V][V] = {{0, 20, 18, 20},
                   {20, 0, 18,12},
                   {18, 18, 0, 6},
                   {10, 12, 6, 0}};

int VISITED_ALL = (1 << V) - 1;

int tsp(int mask,int pos){
    if(mask==VISITED_ALL){
        return graph[pos][0];
    }
    int ans=INT_MAX;
    for(int city=0;city<V;city++){
        if((mask&(1<<city))==0){
            int newAns=graph[pos][city]+tsp(mask|(1<<city),city);
            ans=ans<newAns?ans:newAns;
        }
    }
    return ans;
}

int main(){
    printf("Minimum weight Hamiltonian Circuit: %d\n",tsp(1<<0 ,0));
    return 0;
}
